judge/models/tests/test_profile.py

```diff
@@ -0,0 +1,125 @@
+import base64
+import hmac
+import struct
+
+from django.conf import settings
+from django.test import TestCase
+from django.utils import timezone
+from django.utils.encoding import force_bytes
+
+from judge.models import Profile
+from judge.models.tests.util import CommonDataMixin, create_contest, create_contest_participation
+
+
+class OrganizationTestCase(CommonDataMixin, TestCase):
+    @classmethod
+    def setUpTestData(self):
+        super().setUpTestData()
+        self.profile = self.profiles['normal']
+        self.profile.organizations.add(self.organizations['open'])
+
+    def test_contains(self):
+        self.assertIn(self.profile, self.organizations['open'])
+        self.assertIn(self.profile.id, self.organizations['open'])
+
+        self.assertNotIn(self.profiles['superuser'], self.organizations['open'])
+        self.assertNotIn(self.profiles['superuser'].id, self.organizations['open'])
+
+        with self.assertRaisesRegex(TypeError, 'Organization membership test'):
+            'aaaa' in self.organizations['open']
+
+
+class ProfileTestCase(CommonDataMixin, TestCase):
+    @classmethod
+    def setUpTestData(self):
+        super().setUpTestData()
+        self.profile = self.profiles['normal']
+        self.profile.organizations.add(self.organizations['open'])
+
+    def setUp(self):
+        # We are doing a LOT of field modifications in this test case.
+        # This is to prevent cryptic error messages when a test fails due
+        # to modifications in another test. In theory, no two tests should
+        # touch the same field, but who knows.
+        self.profile.refresh_from_db()
+
+    def test_username(self):
+        self.assertEqual(str(self.profile), self.profile.username)
+
+    def test_organization(self):
+        self.assertIsNone(self.profiles['superuser'].organization)
+        self.assertEqual(self.profile.organization, self.organizations['open'])
+
+    def test_calculate_points(self):
+        self.profile.calculate_points()
+
+        # Test saving
+        for attr in ('points', 'problem_count', 'performance_points'):
+            with self.subTest(attribute=attr):
+                setattr(self.profile, attr, -1000)
+                self.assertEqual(getattr(self.profile, attr), -1000)
+                self.profile.calculate_points()
+                self.assertEqual(getattr(self.profile, attr), 0)
+
+    def test_generate_api_token(self):
+        token = self.profile.generate_api_token()
+
+        self.assertIsInstance(token, str)
+        self.assertIsInstance(self.profile.api_token, str)
+
+        user_id, raw_token = struct.unpack('>I32s', base64.urlsafe_b64decode(token))
+
+        self.assertEqual(self.users['normal'].id, user_id)
+        self.assertEqual(len(raw_token), 32)
+
+        self.assertTrue(
+            hmac.compare_digest(
+                hmac.new(force_bytes(settings.SECRET_KEY), msg=force_bytes(raw_token), digestmod='sha256').hexdigest(),
```

Comment:
I know this also tests that the secret key can actually be made into bytes, but putting this inside of `test_generate_api_token` will throw off debugging. A separate test should probably be made to ensure the secret key can actually be converted into bytes and this test should just use a hard coded string.

Reply:
I'm not sure what you mean by this. What hard coded string should be used? The API token is hashed with the `SECRET_KEY`, so logically, this should also use the `SECRET_KEY` for hashing and comparison.

I also don't see a case of the `SECRET_KEY` ever becoming not convertable to bytes.

Reply:
Right, this is fine. Django [docs](https://docs.djangoproject.com/en/3.0/ref/settings/#std:setting-SECRET_KEY) note that you shouldn't just assume it is, but this could just be telling users not to manually check if it's str or bytes.